### PR TITLE
issue: 784340 Added Flow Tag support

### PR DESCRIPTION
**Pull request by serglelv**
Added Flow Tag functionality support. Please review.
@OphirMunk @alexvm @rosenbaumalex @igor-ivanov @NirNitzani  
Oleg added @olgk 

Signed-off-by: sergeyly sergeyly@mellanox.com

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/573/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/578/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/577/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/582/ for details (Mellanox internal link).

**Comment by serglelv** [comment]
Added FT processing to sockinfo tcp. Please review. 
@OphirMunk @alexvm @rosenbaumalex @olgk @NirNitzani @igor-ivanov

File: configure.ac [hunk 1]
```diff
@@ -352,6 +352,26 @@ else
...
+
+if test "x${with_flow_tag}" = "xyes" ; then
+	AC_COMPILE_CHECK([VERBS Flow Tag check],
+	[#include <infiniband/verbs_exp.h>],
+	[struct ibv_exp_flow_spec_action_tag tag; tag = tag;],
+	ac_cv_libverbs_flow_tag=yes,
+	ac_cv_libverbs_flow_tag=no)
+	AC_MSG_RESULT($ac_cv_libverbs_flow_tag)
+	if test $ac_cv_libverbs_flow_tag = yes ; then
+		AC_DEFINE([FLOW_TAG_ENABLE], [1], [Flow Tagging enable])
```
Comment (olgk) [inline configure.ac]: Please follow the rule on prefixes for definition of experimental VERBS API, see below in the file, should be like IBV_EXP_*. AC_DEFINE looks should have additional prefix DEFINED_

File: src/vma/dev/cq_mgr.cpp [hunk 2]
```diff
@@ -852,8 +863,72 @@ int cq_mgr::vma_poll_and_process_element_rx(mem_buf_desc_t **p_desc_lst)
...
+
+	if (unlikely(m_rx_hot_buff == NULL)) {
+		int index = m_qp->m_mlx5_hw_qp->rq.tail & (m_qp->m_rx_num_wr - 1);
+		m_rx_hot_buff = (mem_buf_desc_t*)(uintptr_t)m_qp->m_rq_wqe_idx_to_wrid[index];
+		m_rx_hot_buff->path.rx.context = NULL;
+		m_rx_hot_buff->path.rx.is_vma_thr = false;
+	}
+	//prefetch_range((uint8_t*)m_rx_hot_buff->p_buffer,safe_mce_sys().rx_prefetch_bytes_before_poll);
+#ifdef RDTSC_MEASURE_RX_VERBS_READY_POLL
+	RDTSC_TAKE_START(g_rdtsc_instr_info_arr[RDTSC_FLOW_RX_VERBS_READY_POLL]);
```
Comment (olgk) [inline src/vma/dev/cq_mgr.cpp]: Alignment?

File: src/vma/dev/ring_bond.h [hunk 3]
```diff
@@ -31,12 +31,15 @@ class ring_bond : public ring {
 	virtual void		mem_buf_desc_completion_with_error_tx(mem_buf_desc_t* p_tx_wc_buf_desc); // Assume locked...
 	virtual void		mem_buf_desc_return_to_owner_rx(mem_buf_desc_t* p_mem_buf_desc, void* pv_fd_ready_array = NULL);
 	virtual void		mem_buf_desc_return_to_owner_tx(mem_buf_desc_t* p_mem_buf_desc);
-	virtual int		get_max_tx_inline();
+	virtual int			get_max_tx_inline();
```
Comment (olgk) [inline src/vma/dev/ring_bond.h]: alignment

File: src/vma/dev/ring_bond.h [hunk 4]
```diff
@@ -31,12 +31,15 @@ class ring_bond : public ring {
...
 	virtual bool		attach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink* sink);
+#if defined(FLOW_TAG_ENABLE)
+	virtual bool		get_flow_tag(uint32_t& tag_id) { tag_id = 0; return false;};
+#endif		
 	virtual bool		detach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink* sink);
 	virtual void		restart(ring_resource_creation_info_t* p_ring_info);
 	virtual mem_buf_desc_t* mem_buf_tx_get(ring_user_id_t id, bool b_block, int n_num_mem_bufs = 1);
-	virtual int		mem_buf_tx_release(mem_buf_desc_t* p_mem_buf_desc_list, bool b_accounting, bool trylock = false);
+	virtual int			mem_buf_tx_release(mem_buf_desc_t* p_mem_buf_desc_list, bool b_accounting, bool trylock = false);
```
Comment (olgk) [inline src/vma/dev/ring_bond.h]: alignment

File: src/vma/dev/ring_simple.cpp [hunk 5]
```diff
@@ -891,29 +1034,24 @@ inline void ring_simple::vma_poll_process_recv_buffer(mem_buf_desc_t* p_rx_wc_bu
...
-
-
-// All CQ wce come here for some basic sanity checks and then are distributed to the correct ring handler
-// Return values: false = Reuse this data buffer & mem_buf_desc
-bool ring_simple::rx_process_buffer(mem_buf_desc_t* p_rx_wc_buf_desc, transport_type_t transport_type, void* pv_fd_ready_array /*=NULL*/)
+#if defined(FLOW_TAG_ENABLE)
+inline void ring_simple::vma_poll_process_recv_buffer_ft(mem_buf_desc_t* p_rx_wc_buf_desc)
 {
-	size_t sz_data = 0;
+	//size_t sz_data = 0;
```
Comment (olgk) [inline src/vma/dev/ring_simple.cpp]: why moved from here? not sure it is allowed to have commented code in production.

File: src/vma/sock/sockinfo.h [hunk 6]
```diff
@@ -132,21 +132,26 @@ class sockinfo : public socket_fd_api, public pkt_rcvr_sink, public pkt_sndr_sou
...
 
-	loops_timer             m_loops_timer;
+	loops_timer		m_loops_timer;
 
 	/**
 	 * list of pending ready packet on the Rx,
 	 * each element is a pointer to the ib_conn_mgr that holds this ready rx datagram
 	 */
-	int			m_n_rx_pkt_ready_list_count;
+	int				m_n_rx_pkt_ready_list_count;
```
Comment (olgk) [inline src/vma/sock/sockinfo.h]: alignment

File: src/vma/dev/ring_simple.h [hunk 7]
```diff
@@ -73,7 +124,13 @@ class ring_simple : public ring
 	void			create_resources(ring_resource_creation_info_t* p_ring_info, bool active) throw (vma_error);
 	// Internal functions. No need for locks mechanism.
 	inline void 		vma_poll_process_recv_buffer(mem_buf_desc_t* p_rx_wc_buf_desc);
+#if defined (FLOW_TAG_ENABLE)
+inline void 		vma_poll_process_recv_buffer_ft(mem_buf_desc_t* p_rx_wc_buf_desc);
```
Comment (olgk) [inline src/vma/dev/ring_simple.h]: alignment in this file should be fixed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/580/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/584/ for details (Mellanox internal link).

**Comment by serglelv** [comment]
Code cleanup. FT define has been changed.
@olgk @OphirMunk

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/584/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/588/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/591/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/595/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/592/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/596/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/597/ for details (Mellanox internal link).

**Comment by NirNitzani** [comment]
handled this issue in #249

**Pull request closed without merging**